lib/rabbitmqctl.ex

```diff
@@ -62,26 +62,65 @@ defmodule RabbitMQCtl do
         {:error, ExitCodes.exit_usage, suggest_message};
       {_, [_|_]} ->
 
-        validation_error({:bad_option, invalid}, command,
-                         unparsed_command, parsed_options);
+        argument_validation_error_output({:bad_option, invalid}, command,
+          unparsed_command, parsed_options);
       _ ->
         options = parsed_options |> merge_all_defaults |> normalize_options
         {arguments, options} = command.merge_defaults(arguments, options)
+
         with_distribution(options, fn() ->
-          validate_and_run_command(options, command, arguments)
-          |> handle_command_output(command, options, unparsed_command, output_fun)
+          # The code below implements a tiny decision tree that has
+          # to do with CLI argument and environment state validation.
+
+          # validate CLI arguments
+          case command.validate(arguments, options) do
+            :ok ->
+              # then optionally validate execution environment
+              case maybe_validate_execution_environment(command, arguments, options) do
+                :ok                        ->
+                  result = proceed_to_execution(command, arguments, options)
+                  handle_command_output(result, command, options, output_fun);
+                {:validation_failure, err} ->
+                  argument_validation_error_output(err, command, unparsed_command, options)
+              end
+            {:validation_failure, err} ->
+              environment_validation_error_output(err, command, unparsed_command, options)
```

Comment:
Should it be `argument_validation_error_output` instead? And `environment_validation_error_output` on line 84?